nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/SnippetResource.java

```diff
@@ -0,0 +1,397 @@
+            populateRemainingSnippetEntityContent(entity);
+            return clusterContext(generateOkResponse(entity)).build();
+        } finally {
+            serviceFacade.cancelRevisions(revisions);
+        }
+    }
+
+    /**
+     * Removes the specified snippet.
+     *
+     * @param httpServletRequest request
+     * @param snippetId The id of the snippet to remove.
+     * @return A entity containing the client id and an updated revision.
+     */
+    @DELETE
+    @Consumes(MediaType.WILDCARD)
+    @Produces(MediaType.APPLICATION_JSON)
+    @Path("{id}")
+    // TODO - @PreAuthorize("hasRole('ROLE_DFM')")
+    @ApiOperation(
+        value = "Deletes the components in a snippet and drops the snippet",
+        response = SnippetEntity.class,
+        authorizations = {
+            @Authorization(value = "Data Flow Manager", type = "ROLE_DFM")
+        }
+    )
+    @ApiResponses(
+        value = {
+            @ApiResponse(code = 400, message = "NiFi was unable to complete the request because it was invalid. The request should not be retried without modification."),
+            @ApiResponse(code = 401, message = "Client could not be authenticated."),
+            @ApiResponse(code = 403, message = "Client is not authorized to make this request."),
+            @ApiResponse(code = 404, message = "The specified resource could not be found."),
+            @ApiResponse(code = 409, message = "The request was valid but NiFi was not in the appropriate state to process it. Retrying the same request later may be successful.")
+        }
+    )
+    public Response deleteSnippet(
+        @Context final HttpServletRequest httpServletRequest,
+        @ApiParam(
+            value = "The snippet id.",
+            required = true
+        )
+        @PathParam("id") final String snippetId) {
+
+        // replicate if cluster manager
+        if (properties.isClusterManager()) {
+            return clusterManager.applyRequest(HttpMethod.DELETE, getAbsolutePath(), getRequestParameters(true), getHeaders()).getResponse();
+        }
+
+        final NiFiUser user = NiFiUserUtils.getNiFiUser();
+
+        // get the revision from this snippet
+        final Set<Revision> revisions = serviceFacade.getRevisionsFromSnippet(snippetId);
+
+        // handle expects request (usually from the cluster manager)
+        final boolean validationPhase = isValidationPhase(httpServletRequest);
+        if (validationPhase || !isTwoPhaseRequest(httpServletRequest)) {
+            // authorize access
+            serviceFacade.authorizeAccess(lookup -> {
+                // ensure read permission to every component in the snippet
+                final Snippet snippet = lookup.getSnippet(snippetId);
+                final Set<Authorizable> authorizables = new HashSet<>();
+                authorizables.addAll(snippet.getProcessGroups().keySet().stream().map(id -> lookup.getProcessGroup(id)).collect(Collectors.toSet()));
```

Comment:
And again, same as above... we should probably actually pull this out into its own method, as it's a decent chunk of code being repeated a few times.